fedimint-client/src/meta.rs

```diff
@@ -0,0 +1,351 @@
+use std::borrow::Cow;
+use std::collections::BTreeMap;
+use std::future::pending;
+use std::time::Duration;
+
+use anyhow::{bail, Context as _};
+use fedimint_core::db::{
+    Committable, Database, DatabaseTransaction, IDatabaseTransactionOpsCoreTyped,
+};
+use fedimint_core::task::waiter::Waiter;
+use fedimint_core::task::{timeout, MaybeSend, MaybeSync};
+use fedimint_core::util::{retry, FibonacciBackoff};
+use fedimint_core::{apply, async_trait_maybe_send};
+use futures::future::join_all;
+use futures::StreamExt;
+use serde::de::DeserializeOwned;
+use tracing::{debug, instrument, warn};
+
+use crate::db::{MetaFieldKey, MetaFieldName, MetaFieldPrefix, MetaFieldValue};
+use crate::Client;
+
+/// Tries to parse `str_value` as JSON. In the special case that `V` is `String`
+/// we return the raw `str_value` if JSON parsing fails. This necessary since
+/// the spec wasn't clear enough in the beginning.
+#[instrument(err)] // log on every failure
+pub fn parse_meta_value_static<V: DeserializeOwned + 'static>(
+    str_value: &str,
+) -> anyhow::Result<V> {
+    let res = serde_json::from_str(str_value)
+        .with_context(|| format!("Decoding meta field value '{str_value}' failed"));
+
+    // In the past we encoded some string fields as "just a string" without quotes,
+    // this code ensures that old meta values still parse since config is hard to
+    // change
+    if res.is_err() && std::any::TypeId::of::<V>() == std::any::TypeId::of::<String>() {
+        let string_ret = Box::new(str_value.to_owned());
+        let ret: Box<V> = unsafe {
+            // We can transmute a String to V because we know that V==String
+            std::mem::transmute(string_ret)
+        };
+        Ok(*ret)
+    } else {
+        res
+    }
+}
+
+#[apply(async_trait_maybe_send!)]
+pub trait MetaSource: MaybeSend + MaybeSync {
+    /// This must be stable.
+    fn source_id(&self) -> u64;
+    /// Wait for next change in this source.
+    async fn wait_for_update(&self);
+    async fn fetch(
+        &self,
+        client: &Client,
+    ) -> anyhow::Result<BTreeMap<MetaFieldName, MetaFieldValue>>;
+}
+
+/// Service for managing the caching of meta fields.
+#[derive(Default)]
+pub struct MetaService {
+    sources: Vec<Box<dyn MetaSource>>,
+    initial_fetch_waiter: Waiter,
+}
+
+#[derive(Clone, Debug, Copy)]
+pub struct TypedMetaValue<T> {
+    pub value: T,
+    /// None means unknown revision
+    pub revision: Option<u64>,
+    pub priority: u64,
+}
+
+impl MetaService {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub async fn get_field<V: DeserializeOwned + 'static>(
```

Comment:
We should document that this can block for a significant amount of time on first run (currently for around 60s).

Reply:
fixed